HW1/P6/P6.py

```diff
@@ -0,0 +1,130 @@
+import pyspark
+from pyspark import SparkContext
+import time
+import random
+from random import randint
+
+f = open('P6.txt', 'w')
+
+def isNotNumber(word):
+	try:
+		int(word)
+		return False
+	except ValueError:
+		return True
+
+#take a list and then return a list
+#of (word, num times word appears in list)
+
+def listToCounts(lst):
+	results = []
+	dic = {}
+	for word in lst:
+		if word in dic:
+			dic[word] += 1
+		else:
+			dic[word] = 1
+	for word in lst:
+		results.append((word, dic[word]))
+	return list(set(results))
+
+#Determine which list is the word-pair
+#and set the other one to be the listed "third word"
+#Used in a map function later. 
+def incorporateThirdWord(x, y):
+	if len(x) == 2:
+		return (x, [y])
+	else: 
+		return (y, [x])
+
+sc = SparkContext()
+sc.setLogLevel('ERROR')
+#sc.parallelize('s3://Harvard-CS205/Shakespeare/Shakespeare.txt')
+rawText = sc.textFile('Shakespeare.txt')
+
+#Clean the data to make sure it doesn't have the concerns listed
+#in the problem statement. 
+words = rawText.flatMap(lambda x: x.split(' '))
+words = words.filter(lambda x: len(x) > 0)
+noNumbers = words.filter(lambda x: False if x.isdigit() else True)
+noCaps = noNumbers.filter(lambda x: False if x.isupper() else True)
+noCapsPeriod = noCaps.filter(lambda x: False if (x[:len(x)-1].isupper() and x[-1] is '.') else True)
+
+#index these words so we can go ahead and get a feeling of which ones come first
+#or second.
+text = noCapsPeriod.zipWithIndex()
+
+#The text, normally with the word first, and then the index of the word as the value
+text = text.map(lambda x: (x[1], (x[0])))
+#Shift all these back one. So now, index "n" in text is the "nth" word in Shakespeare
+#and #n# in one_shift is the "n+1"th word in Shakespeare
+one_shift = text.map(lambda (x, y): (x-1, y))
+two_shift = text.map(lambda (x, y): (x-2, y)).sortByKey()
+
+#Join and reduce by key so we can get pairs of words. For some reason
+#Reduce by keys messes with the order, so sort it afterwards. 
```

Comment:
It might have been the join...

http://stackoverflow.com/questions/29284095/which-operations-preserve-rdd-order